crates/worker/src/partition/state_machine/entries/notification.rs

```diff
@@ -7,37 +7,151 @@
 // As of the Change Date specified in that file, in accordance with
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
+use tracing::trace;
 
-use crate::partition::state_machine::lifecycle::ResumeInvocationCommand;
-use crate::partition::state_machine::{CommandHandler, Error, StateMachineApplyContext};
-use restate_storage_api::invocation_status_table::InvocationStatus;
+use restate_service_protocol_v4::entry_codec::ServiceProtocolV4Codec;
+use restate_storage_api::invocation_status_table::{InFlightInvocationMetadata, InvocationStatus};
 use restate_storage_api::journal_table_v2::ReadOnlyJournalTable;
+use restate_tracing_instrumentation as instrumentation;
 use restate_types::identifiers::InvocationId;
-use restate_types::journal_v2::NotificationId;
 use restate_types::journal_v2::raw::RawNotification;
+use restate_types::journal_v2::{Command, CommandType, CompletionId, NotificationId};
+
+use crate::partition::state_machine::lifecycle::ResumeInvocationCommand;
+use crate::partition::state_machine::{CommandHandler, Error, StateMachineApplyContext};
 
 pub(super) struct ApplyNotificationCommand<'e> {
     pub(super) invocation_id: InvocationId,
     pub(super) invocation_status: &'e mut InvocationStatus,
     pub(super) entry: &'e RawNotification,
 }
 
+impl<'e> ApplyNotificationCommand<'e> {
+    async fn create_trace_span<'s, S>(
+        &self,
+        ctx: &'_ mut StateMachineApplyContext<'s, S>,
+        completion_id: CompletionId,
+        invocation_metadata: &InFlightInvocationMetadata,
+    ) -> Result<(), Error>
+    where
+        S: ReadOnlyJournalTable,
+    {
+        let (header, command) = ctx.storage
+                    .get_command_by_completion_id(self.invocation_id, completion_id)
+                    .await?.expect("For given completion id, the corresponding command must be present already in the journal");
+
+        if !instrumentation::is_service_tracing_enabled()
+            || !matches!(
+                command.command_type(),
+                CommandType::Run | CommandType::Sleep | CommandType::GetPromise
+            )
```

Comment:
I think you can also add to this condition checking if the parent span is being sampled:

```
invocation_metadata.journal_metadata.span_context.is_sampling()
```